subprojects/docs/src/docs/userguide/developing-plugins/kotlin_gradle_plugins.adoc

```diff
@@ -0,0 +1,334 @@
+= Writing Gradle plugins in Kotlin
+
+In addition to <<kotlin_dsl.adoc, writing build scripts>>, Gradle also supports writing plugins in https://www.jetbrains.com/opensource/kotlin/[Kotlin].
+
+There are multiple options to write your Kotlin plugins and multiple compatibility considerations. This page is a non-exhaustive list of dependencies and compiler configurations and their impacts in terms of compatibility and usability.
+
+It also describes the different development plugins to help you set up these different options.
+
+
+[[sec:dependencies]]
+== Different dependencies
+
+Like other JVM projects, Gradle plugins are jar files with dependencies. However,  unlike other JVM projects, Gradle provides fixed versions of some dependencies at runtime. Those dependencies are `compileOnly` dependencies and need to be compatible with the version of Gradle your plugin is running in.
+
+If you are writing convention plugins, this is usually not an issue. But if you distribute your plugins, you should pay attention to their compatibility.
+
+[[sec:gradle_api]]
+=== The Gradle API
+
+All plugins use the https://docs.gradle.org/current/javadoc/index.html[Gradle Core API] to configure a Gradle build. The Gradle API must be added as a build compile-time dependency using the helper function `gradleApi()`
+
+[source,kotlin]
+----
+dependencies {
+    compileOnly(gradleApi())
+}
+----
+
+TIP: `gradleApi()` must be a `compileOnly` dependency, because the correct Gradle API will be provided by Gradle at runtime.
```

Comment:
This somehow contradicts with the current documentation which suggest declaring the dependency as `implementation`: https://docs.gradle.org/current/userguide/dependency_management.html#sub:api_dependencies

Do you think the documentation on `gradleApi()` and `localGroovy()` should be fixed to `compileOnly` as well?

Reply:
Good find. I would say so but I'll defer to the Gradle team on this. Maybe there's a reason it's like this.